Source API

rodjek opened the pull request:
Opening this to discuss how I think the Source API should look.

What I really want to prevent is sources turning into implementation clusterfucks ala librarian.  I'd like the Sources to only define a handful of required public methods:
- `fetch!`
- `versions`
- `dependencies` (maybe?  if we only fetch dependencies from the metadata files in the fetched modules we can possibly even abstract this away into `Henson::Source::Generic`

Then in `Henson::Source::Generic` we have the meat of the process
- `fetch_location` - If caching enabled, returns the cache location for the module, otherwise a temporary path
- `install!` - Install the module from `fetch_location` to the final destination
- etc

Basically, I want the Source objects to be as simple as possible - grab data and put it into a standard location (module contents -> standard path on disk, metadata -> instance vars).

rodjek committed:
Remove corner case for Path source

Always call #fetch! on the source object regardless of the source type

rodjek committed:
Add stub fetch! methods to sources

rodjek committed:
Change implementation methods for Path source to private methods

rodjek committed:
All Sources should implement #versions

rodjek committed:
Add method to Path source to parse version from the Modulefile

rodjek committed:
Basic #versions implementation for Path source

rodjek committed:
logic to check if fetched version meets the module requirement

wfarr commented on lib/henson.rb:
```diff
@@ -28,4 +28,6 @@ def self.exit_code(i)
   class PuppetfileError    < Error; exit_code(14); end
   class PuppetfileNotFound < Error; exit_code(16); end
+  class ModuleNotFound     < Error; exit_code(18); end
+  class RequirementNotSatisfied < Error; exit_code(20); end
```
Alignment, because I'm picky.

wfarr commented on lib/henson/dsl.rb:
```diff
@@ -26,10 +26,13 @@ def evaluate(puppetfile)
+        unless puppet_module.satisfied?
+          raise RequirementNotSatisfied, "#{name} (#{version})"
+        end
+        puppet_module.source.fetch!
```
I really feel that `puppet_module` ought to own fetch here, which can delegate to the source's fetch because Law of Demeter and all that.

rodjek replied:
Agreed

wfarr commented on lib/henson/source/path.rb:
```diff
@@ -14,6 +24,25 @@ def valid?
         path && File.directory?(path)
       end
+
+      def version_from_modulefile
```
I'm kind of contemplating adding multiple DSL classes, one of which would be a `DSL::Modulefile`. In this case, we could just define `version` in that DSL and then set `method_missing` there to just silently ignore any other bits.

rodjek replied:
I'm down with that, this was just a quick hack to show my intentions.

wfarr replied:
Rad, can always refactor later. Please drop a `TODO` comment to that effect.

wfarr commented on spec/henson/source/path_spec.rb:
```diff
@@ -16,24 +16,31 @@
     it "returns true if path_exists? is true" do
       instance.stubs(:path_exists?).returns(true)
-      instance.valid?.should be_true
+      instance.send(:valid?).should be_true
```
Why all the `send`s here?

rodjek replied:
Changed them to be private methods to make the public interface of the sources cleaner

wfarr replied:
Dig it. There's other places in the code that could likely use that treatment as well.

wfarr commented:
:+1: on light sources — totally my intent also. A Source should only know:
- what versions are available
- if the version requested is currently installed
- how to install a version
- how to cache an installed version

---

Dependency lookup should be managed by `Source::Generic`. We should have a cascade of preferred dependency places we look at:
1. a `Modulefile`
2. a `Puppetfile`

These will both need to leverage DSL classes specific to their file.

rodjek commented:
> Dependency lookup should be managed by `Source::Generic`

Totally agree.  I sketched out that method in `Source::Path` but it felt wrong to have it there.

wfarr commented:
Rad.

wfarr commented:
Merge/rebase master to get the simple delegator lulz.

wfarr committed:
Merge origin/master

wfarr committed:
Code coverage for Source::Path

wfarr committed:
Fix coverage for friendly errors

wfarr committed:
Fix coverage for Source::Generic

wfarr committed:
Make Source responsible for determining requirement satisfaction

wfarr committed:
Fix indentation

wfarr committed:
Fix coverage for satisfication of module requirements

wfarr committed:
Remove simple delegator gem

Not truly necessary to save 2 lines of typing per delegated method.

wfarr committed:
add rake todo

wfarr committed:
Add ModulefileNotFound error

wfarr committed:
Make DSL::Modulefile raise an error if no version is defined in the Modulefile

wfarr committed:
Move ModulefileNotFound logic into DSL::Modulefile

wfarr committed:
Move ModuleNotFound checking on Source::Path into constructor

wfarr committed:
Delegate PuppetModule.versions to its source

wfarr commented:
This is now in-line with master and I've (obviously) refactored a bunch of stuff.

I really want to knuckle down and "finish" this PR and merge it into master.

Here's what I think is remaining/want to get done to merge this:
- Eliminate `Source::Path#version_from_modulefile` and instead use the DSL inferred version
- Make dependencies on `DSL::Modulefile` into `PuppetModule`s instead of hashes.
- Make `Installer` check for Modulefile first, then Puppetfile, and otherwise raise an error about no metadata source.

From there we can do smaller PRs for remaining things on master, otherwise this PR will just go epic and be confusing.

wfarr committed:
Change rspec formatter to dots

wfarr committed:
Installer calls fetch! for each module

Interim hack just to try and get some basic output — this won't suffice
long-term

wfarr committed:
Move module fetching back to the DSL for now

wfarr committed:
Don't make Puppetfile::DSL fetch the modules

Installer really does need to handle this because we'll use the
Puppetfile::DSL for other things also.

wfarr committed:
Use the Modulefile DSL to infer version with Source::Path

wfarr committed:
Add path setting

wfarr committed:
Delege PuppetModule's install! to the source

wfarr committed:
Add path flag to install

wfarr committed:
Installer makes path, fetches, and installs modules

wfarr committed:
Ignore new shared directories created by tests

wfarr committed:
Add forge support to the Puppetfile DSL

wfarr committed:
All sources now take the module name as the first arg

wfarr committed:
CLI fixes

wfarr committed:
Move source installed? to Generic

wfarr committed:
Add basic log output for installing

wfarr committed:
Add helpers for Installer to PuppetModule

wfarr committed:
Add specs for default version on a Source::Path if no Modulefile

rodjek commented on lib/henson/installer.rb:
```diff
@@ -1,9 +1,22 @@
+      FileUtils.mkdir_p File.expand_path(Henson.settings[:path])
+
+      parse_puppetfile!.modules.each do |mod|
+        mod.fetch! if mod.needs_fetching?
```
We'll probably need to split this out into two loops, one to fetch and then one to install as we'll eventually have a dependency resolution phase that'll have to happen after fetch but before install.

wfarr replied:
Probably right.  

On Saturday, November 3, 2012 at 7:58 PM, Tim Sharpe wrote:

> In lib/henson/installer.rb:
> 
> > > module Henson > class Installer > def self.install! > - parse_puppetfile! > + FileUtils.mkdir_p File.expand_path(Henson.settings[:path]) > + > + parse_puppetfile!.modules.each do |mod| > + mod.fetch! if mod.needs_fetching?  
> > > We'll probably need to split this out into two loops, one to fetch and then one to install as we'll eventually have a dependency resolution phase that'll have to happen after fetch but before install.
> 
> —
> Reply to this email directly or view it on GitHub (https://github.com/wfarr/henson/pull/1/files#r2024398).

rodjek commented:
Looks good, I say we merge this behemoth :)

wfarr commented:
Merging.

Then let's focus on getting the path source support with transitive dependencies working (assuming path sources only on both sides), and then implementing the same sauce for the git source.
